sample_apps/rust/examples/utils/query_common.rs

```diff
@@ -0,0 +1,261 @@
+    pub region: String,
+
+    // The Timestream for LiveAnalytics table name to use for all queries
+    #[arg(short, long, default_value = DEFAULT_TABLE_NAME)]
+    pub table_name: String,
+}
+
+pub async fn get_connection(
+    region: &String,
+) -> Result<timestream_query::Client, timestream_query::Error> {
+    let config = aws_config::defaults(aws_config::BehaviorVersion::latest())
+        .region(Region::new(region.to_owned()))
+        .load()
+        .await;
+    let (client, reload) = timestream_query::Client::new(&config)
+        .with_endpoint_discovery_enabled()
+        .await
+        .expect("Failure to get the query client connection with Timestream");
+    tokio::task::spawn(reload.reload_task());
+    Ok(client)
+}
+
+pub fn write(mut file: &fs::File, s: String) -> Result<(), Box<dyn Error>> {
+    let s_formatted = format!("{}\n", s);
+    if file.write(s_formatted.as_bytes()).is_err() {
+        return Err(String::from("Failed to write to file").into());
+    }
+    if file.flush().is_err() {
+        return Err(String::from("Failed to flush file").into());
+    }
+    Ok(())
+}
+
+#[allow(dead_code)]
+pub fn process_scalar_type(data: &types::Datum) -> Result<String, Box<dyn Error>> {
+    data.scalar_value
+        .clone()
+        .ok_or("Scalar value is None".to_string().into())
+}
+
+#[allow(dead_code)]
+pub fn process_time_series_type(
+    data: &[types::TimeSeriesDataPoint],
+    column_info: &types::ColumnInfo,
+) -> Result<String, Box<dyn Error>> {
+    let mut value = String::new();
+    for (i, datum) in data.iter().enumerate() {
+        value.push_str(&datum.time);
+        value.push(':');
+
+        let column_type = column_info.r#type();
+        let column_type_ref = column_type.as_ref().ok_or("Column type is None")?;
+        let scalar_type = column_type_ref.scalar_type.to_owned();
+        let scalar_type_ref = scalar_type.ok_or("Scalar type is None")?;
+
+        let datum_value = datum.value.as_ref().ok_or("Datum value is None")?;
+        let _scalar_value = datum_value
+            .scalar_value
+            .to_owned()
+            .ok_or("Scalar value is None")?;
+
+        if scalar_type_ref.as_str() == "" {
+            value.push_str(&process_scalar_type(datum_value)?);
+        } else if let Some(array_column_info) = &column_type_ref.array_column_info {
+            let array_value = datum_value
+                .array_value
+                .as_ref()
+                .ok_or("Array value is None")?;
+            value.push_str(&process_array_type(array_value, array_column_info)?);
+        } else if let Some(row_column_info) = &column_type_ref.row_column_info {
+            let row_value = datum_value.row_value.as_ref().ok_or("Row value is None")?;
+            value.push_str(&process_row_type(&row_value.data, row_column_info)?);
+        } else {
+            panic!("Bad data type");
+        }
+
+        if i != data.len() - 1 {
+            value.push_str(", ");
+        }
```

Comment:
You can avoid the index math here by rewriting this using an iterator (you already have one from enumerate you could map or fold over), collecting into a `Vec` and then using https://doc.rust-lang.org/std/vec/struct.Vec.html#method.join

This can be done without the additional alloc using the `itertools` crate, but I imagine it's not that big of a deal for a sample app.

Reply:
applies throughout

Reply:
I have changed all of the type processing functions to use `map` and joining. I have also fixed an issue where arrays and rows weren't being processed.